### PR TITLE
[SPARK-24097][ML] Instrumentation improvements - RandomForest and GradientBoostedTree

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes were proposed in this pull request?

Instruments improvements for `RandomForest` and `GradientBoostedTree` in `ml.tree.impl` package.
**Note:** I add `Instrumentation` and `OptionalInstrumentation` with `Serializable`. Because tree algos need to ouput some logs on executor side and the instrumentation object need to be broadcast.

## How was this patch tested?

Manual.

### CODE REVIEW

**Comment by jkbradley** [comment]
Regarding logging on executors, are you OK with the proposed plan here?  https://issues.apache.org/jira/browse/SPARK-23686
Basically, we'll keep using regular Logging on executors, rather than using Instrumentation on them.

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala [hunk 1]
```diff
@@ -22,12 +22,11 @@ import java.io.IOException
 import scala.collection.mutable
 import scala.util.Random
 
-import org.apache.spark.internal.Logging
 import org.apache.spark.ml.classification.DecisionTreeClassificationModel
 import org.apache.spark.ml.feature.LabeledPoint
 import org.apache.spark.ml.regression.DecisionTreeRegressionModel
 import org.apache.spark.ml.tree._
-import org.apache.spark.ml.util.Instrumentation
+import org.apache.spark.ml.util.{Instrumentation, OptionalInstrumentation}
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala]: Instrumentation is no longer needed

**Review by jkbradley — COMMENTED** [review]
Thanks for the PR!

Can you please change the PR title to use "Instrumentation" instead of "Instruments"?  It's nice to use consistent terminology since it makes things much easier to search for later on.

For GBTs, I'd recommend passing the Instrumentation instance created in each Estimator down to the shared implementation.  That way, logging will have the Estimator's exact UID.

**Comment by WeichenXu123** [comment]
Jenkins, test this please.

**Comment by jkbradley** [comment]
LGTM pending fresh tests
Thanks!

**Comment by jkbradley** [comment]
jenkins test this please

**Comment by WeichenXu123** [comment]
Jenkins, test this please.

**Pull request closed without merging**